pkg/authorization/admission/restrictusers/subjectchecker.go

```diff
@@ -0,0 +1,247 @@
+type roleBindingRestrictionContext struct {
+	oclient oclient.Interface
+	kclient kclientset.Interface
+
+	// namespace is the namespace for which the RoleBindingRestriction makes
+	// determinations.
+	namespace string
+}
+
+// newRoleBindingRestrictionContext returns a new roleBindingRestrictionContext
+// object.
+func newRoleBindingRestrictionContext(ns string, kc kclientset.Interface,
+	oc oclient.Interface) (*roleBindingRestrictionContext, error) {
+	return &roleBindingRestrictionContext{
+		namespace: ns,
+		kclient:   kc,
+		oclient:   oc,
+	}, nil
+}
+
+// aggregatedRoleBindingRestrictions represents the union of zero or more
+// RoleBindingRestriction objects.
+type aggregatedRoleBindingRestrictions struct {
+	// ctx holds context that is required to make a determination whether
+	// a restriction applies to a subject.
+	ctx *roleBindingRestrictionContext
+
+	// userRestrictions match against user subjects.
+	userRestrictions []*authorizationapi.UserRestriction
+
+	// groupRestrictions match against group subjects.
+	groupRestrictions []*authorizationapi.GroupRestriction
+
+	// serviceAccountRestrictions match against service-account subjects.
+	serviceAccountRestrictions []*authorizationapi.ServiceAccountRestriction
+}
+
+// newAggregatedRoleBindingRestrictions returns a new
+// aggregatedRoleBindingRestrictions object.
+func newAggregatedRoleBindingRestrictions(ctx *roleBindingRestrictionContext) (*aggregatedRoleBindingRestrictions, error) {
+	return &aggregatedRoleBindingRestrictions{ctx: ctx}, nil
+}
+
+func (r *aggregatedRoleBindingRestrictions) addRolebindingRestriction(rbr *authorizationapi.RoleBindingRestriction) {
+	switch {
+	case rbr.Spec.UserRestriction != nil:
+		r.userRestrictions = append(r.userRestrictions, rbr.Spec.UserRestriction)
+	case rbr.Spec.GroupRestriction != nil:
+		r.groupRestrictions = append(r.groupRestrictions, rbr.Spec.GroupRestriction)
+	case rbr.Spec.ServiceAccountRestriction != nil:
+		r.serviceAccountRestrictions = append(r.serviceAccountRestrictions, rbr.Spec.ServiceAccountRestriction)
+	}
+}
+
+// Allowed makes the determination whether the aggregatedRoleBindingRestrictions
+// allow rolebindings on the given subject.
+func (r *aggregatedRoleBindingRestrictions) Allowed(subject *kapi.ObjectReference) error {
+	switch subject.Kind {
+	case authorizationapi.SystemUserKind:
+		for _, userRestriction := range r.userRestrictions {
+			allowed, err := systemUserAllowed(subject.Name, userRestriction)
+			if err != nil {
+				return err
+			}
+			if allowed {
+				return nil
+			}
+		}
+
+	case authorizationapi.SystemGroupKind:
+		for _, groupRestriction := range r.groupRestrictions {
+			allowed, err := systemGroupAllowed(subject.Name, groupRestriction)
+			if err != nil {
+				return err
+			}
+			if allowed {
+				return nil
+			}
+		}
+
+	case authorizationapi.UserKind:
+		user, err := r.ctx.oclient.Users().Get(subject.Name)
```

Comment:
Try to match literals before checking for label matches.  We want to allow pre-binding if possible.  Similarly, for user matching, try to use the userToGroup cache to pull the group information without requiring a user be present ahead of time.